pop_tools/zonal_average.py

```diff
@@ -0,0 +1,247 @@
+        self.dx = dx
+        self.dy = dy
+
+        # Set the regridding method
+        self.regrid_method = regrid_method
+
+        # Set the grid generation method
+        self.method_gen_grid = method_gen_grid
+
+        # If the user does not input a mask, use default mask
+        if not mask:
+            self.mask = self.grid['REGION_MASK']
+
+            if 'region_name' not in self.grid.variables:
+                self.mask_labels = np.unique(self.mask.values)
+
+            else:
+                self.mask_labels = self.grid['region_name']
+
+        else:
+            self.mask = mask
+
+    # Setup method for regridding a dataarray
+    def _regrid_dataarray(self, da_in, regrid_mask=False, regrid_method=None):
+
+        src_grid = _convert_to_xesmf(da_in, self.grid)
+        dst_grid = _generate_dest_grid(self.dy, self.dx, self.method_gen_grid)
+
+        # If the user does not specify a regridding method, use default conservative
+        if regrid_method is None:
+            regridder = _generate_weights(src_grid, dst_grid, self.regrid_method)
+
+        else:
+            regridder = _generate_weights(src_grid, dst_grid, regrid_method)
+
+        # Regrid the input data array, assigning the original attributes
+        da_out = regridder(src_grid[da_in.name])
+        da_out.attrs = da_in.attrs
+
+        return da_out
+
+    def regrid(self, obj, **kwargs):
+        """generic interface for regridding DataArray or Dataset"""
+        if isinstance(obj, xr.Dataset):
+            ds_list = []
+            for var in obj.variables:
+
+                # Make sure the variable has the correct dimensions, is not a coordinate, and is not a velocity
+                if (
+                    ('nlat' and 'nlon' in obj[var].dims)
+                    and (var not in list(obj.coords))
+                    and ('velocity' not in obj[var].long_name.lower())
+                ):
+                    ds_list.append(obj[var])
+            return xr.merge(ds_list).map(self._regrid_dataarray, keep_attrs=True, **kwargs)
+        elif isinstance(obj, xr.DataArray):
+            return self._regrid_dataarray(obj, **kwargs)
+        raise TypeError('input data must be xarray DataArray or xarray Dataset!')
+
+    def zonal_average(self, obj, vertical_average=False, **kwargs):
+
+        data = self.regrid(obj, **kwargs)
+        mask = self.regrid(self.mask, regrid_method='nearest_s2d', **kwargs)
+
+        # Store the various datasets seperated by basin in this list
+        ds_list = []
+        for region in tqdm(np.unique(mask)):
+
+            if region != 0:
+                ds_list.append(data.where(mask == region).groupby('lat').mean())
+
+        # Merge the datasets
+        out = xr.concat(ds_list, dim='nreg')
+
+        # Check to see if a weighted vertical average is needed
+        if vertical_average:
+
+            # Run the vertical, weighted average
+            out = out.weighted(out['z_t'].fillna(0)).mean(dim=['z_t'])
```

Comment:
why weight by `z_t` and not `DZT`?

Reply:
Could this be calculated by using the following?

```python
DZT = out.z_t.diff(dim='z_t')
```

Or am I misunderstanding what DZT is?

Reply:
`out.z_t.diff(dim='z_t')` will return the distance between layer centers, and for 60 levels it will only return an array of length 59:

```py3
>>> import pop_tools
>>> gx1v7_grid = pop_tools.get_grid("POP_gx1v7")
>>> gx1v7_grid.z_t.diff(dim='z_t')
<xarray.DataArray 'z_t' (z_t: 59)>
array([ 1000.     ,  1000.     ,  1000.     ,  1000.     ,  1000.     ,
        1000.     ,  1000.     ,  1000.     ,  1000.     ,  1000.     ,
        1000.     ,  1000.     ,  1000.     ,  1000.     ,  1000.     ,
        1009.8404 ,  1038.0646 ,  1081.22175,  1136.90105,  1205.11015,
        1286.69055,  1383.0544 ,  1496.13345,  1628.40275,  1782.946  ,
        1963.55735,  2174.8772 ,  2422.5496 ,  2713.41105,  3055.7061 ,
        3459.30485,  3935.90165,  4499.1272 ,  5164.489  ,  5948.97315,
        6870.06835,  7943.9187 ,  9182.2754 , 10588.062  , 12149.8325 ,
       13836.2715 , 15593.0264 , 17344.9219 , 19005.6846 , 20494.0957 ,
       21751.2334 , 22751.50195, 23502.97165, 24038.333  , 24401.99805,
       24638.8965 , 24787.6709 , 24878.15135, 24931.6328 , 24962.4424 ,
       24979.77735, 24989.31735, 24994.45895, 24997.17675])
>>>
```

You want the cell thickness - the difference between layer interfaces, which is  `z_w_bot - z_w`. Those arrays should be the same length although the dimension names are different, but here's a kludgy example of at least returning a DataArray (the coordinates need to be cleaned up):

```py3
>>> gx1v7_grid.z_w_bot.swap_dims(z_w_bot="z_t") - gx1v7_grid.z_w.swap_dims(z_w="z_t")
<xarray.DataArray (z_t: 60)>
array([ 1000.    ,  1000.    ,  1000.    ,  1000.    ,  1000.    ,
        1000.    ,  1000.    ,  1000.    ,  1000.    ,  1000.    ,
        1000.    ,  1000.    ,  1000.    ,  1000.    ,  1000.    ,
        1000.    ,  1019.6808,  1056.4484,  1105.9951,  1167.807 ,
        1242.4133,  1330.9678,  1435.141 ,  1557.1259,  1699.6796,
        1866.2124,  2060.9023,  2288.8521,  2556.2471,  2870.575 ,
        3240.8372,  3677.7725,  4194.0308,  4804.2236,  5524.7544,
        6373.1919,  7366.9448,  8520.8926,  9843.6582, 11332.4658,
       12967.1992, 14705.3438, 16480.709 , 18209.1348, 19802.2344,
       21185.957 , 22316.5098, 23186.4941, 23819.4492, 24257.2168,
       24546.7793, 24731.0137, 24844.3281, 24911.9746, 24951.291 ,
       24973.5938, 24985.9609, 24992.6738, 24996.2441, 24998.1094])
Coordinates:
    z_w_bot  (z_t) float64 1e+03 2e+03 3e+03 4e+03 ... 5e+05 5.25e+05 5.5e+05
    z_w      (z_t) float64 0.0 1e+03 2e+03 3e+03 ... 4.75e+05 5e+05 5.25e+05
Dimensions without coordinates: z_t
>>>
```

I think `out` might already have a `dz` data array, though, which would be the best thing to use:

```py3
>>> gx1v7_grid["dz"]
<xarray.DataArray 'dz' (z_t: 60)>
array([ 1000.    ,  1000.    ,  1000.    ,  1000.    ,  1000.    ,
        1000.    ,  1000.    ,  1000.    ,  1000.    ,  1000.    ,
        1000.    ,  1000.    ,  1000.    ,  1000.    ,  1000.    ,
        1000.    ,  1019.6808,  1056.4484,  1105.9951,  1167.807 ,
        1242.4133,  1330.9678,  1435.141 ,  1557.1259,  1699.6796,
        1866.2124,  2060.9023,  2288.8521,  2556.2471,  2870.575 ,
        3240.8372,  3677.7725,  4194.0308,  4804.2236,  5524.7544,
        6373.1919,  7366.9448,  8520.8926,  9843.6582, 11332.4658,
       12967.1992, 14705.3438, 16480.709 , 18209.1348, 19802.2344,
       21185.957 , 22316.5098, 23186.4941, 23819.4492, 24257.2168,
       24546.7793, 24731.0137, 24844.3281, 24911.9746, 24951.291 ,
       24973.5938, 24985.9609, 24992.6738, 24996.2441, 24998.1094])
Coordinates:
  * z_t      (z_t) float64 500.0 1.5e+03 2.5e+03 ... 5.125e+05 5.375e+05
Attributes:
    units:      cm
    long_name:  thickness of layer k
>>>
```

Reply:
I adjusted this to utilize the `dz` dataarray within the grid
```python
out = out.weighted(self.grid.dz).mean(dim='z_t')
```
Which seems to work well.